⚠️ MachinePool API Controller Implementation

juan-lee opened the pull request:
**What this PR does / why we need it**:
Adds `MachinePool` controller implementation described in the [MachinePool API proposal](https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/20190919-machinepool-api.md).

juan-lee commented:
FYI - I'll add tests before removing the [WIP] flag.

ncdc commented:
@juan-lee is this ready for an initial round of reviews?

juan-lee commented:
> @juan-lee is this ready for an initial round of reviews?

Yes, please. @ncdc

CecileRobertMichon commented on controllers/machinepool_controller.go:
```diff
@@ -57,13 +76,181 @@ func (r *MachinePoolReconciler) SetupWithManager(mgr ctrl.Manager, options contr
+	return r.reconcile(ctx, cluster, mp)
+}
+
+func (r *MachinePoolReconciler) reconcile(ctx context.Context, cluster *clusterv1.Cluster, mp *clusterv1.MachinePool) (ctrl.Result, error) {
```
let's add unit tests for these new controller files

vincepri commented on controllers/machinepool_controller.go:
```diff
@@ -57,13 +76,181 @@ func (r *MachinePoolReconciler) SetupWithManager(mgr ctrl.Manager, options contr
+	}
+	mp.Labels[clusterv1.ClusterLabelName] = mp.Spec.ClusterName
+
+	cluster, err := util.GetClusterByName(ctx, r.Client, mp.ObjectMeta.Namespace, mp.Spec.ClusterName)
```
```suggestion
	cluster, err := util.GetClusterByName(ctx, r.Client, mp.Namespace, mp.Spec.ClusterName)
```

vincepri commented on controllers/machinepool_controller.go:
```diff
@@ -57,13 +76,181 @@ func (r *MachinePoolReconciler) SetupWithManager(mgr ctrl.Manager, options contr
+		if err := patchHelper.Patch(ctx, mp); err != nil {
+			if reterr == nil {
+				reterr = err
+			}
```
We can use this pattern to aggregate errors:
`reterr = kerrors.NewAggregate([]error{reterr, err})`

vincepri replied:
without the need for the `if reterr == nil`

vincepri commented on controllers/machinepool_controller.go:
```diff
@@ -57,13 +76,181 @@ func (r *MachinePoolReconciler) SetupWithManager(mgr ctrl.Manager, options contr
+
+func (r *MachinePoolReconciler) reconcileDelete(ctx context.Context, cluster *clusterv1.Cluster, mp *clusterv1.MachinePool) (ctrl.Result, error) {
+	if mp.Spec.ProviderIDs == nil {
+		return ctrl.Result{}, errNilProviderID
```
Do we actually need to guard against a nil list here? What if the list never gets initialized, and we issue a delete? Wouldn't the delete get stuck because of the finalizer?

vincepri commented on controllers/machinepool_controller_phases.go:
```diff
@@ -0,0 +1,232 @@
+
+// reconcileBootstrap reconciles the Spec.Bootstrap.ConfigRef object on a MachinePool.
+func (r *MachinePoolReconciler) reconcileBootstrap(ctx context.Context, mp *clusterv1.MachinePool) error {
+	// TODO(vincepri): Move this validation in kubebuilder / webhook.
```
We now have webhooks in place :)

vincepri commented on controllers/machinepool_controller_phases.go:
```diff
@@ -0,0 +1,232 @@
+	// TODO(vincepri): Move this validation in kubebuilder / webhook.
+	if mp.Spec.Template.Spec.Bootstrap.ConfigRef == nil && mp.Spec.Template.Spec.Bootstrap.Data == nil {
+		return errors.Errorf(
+			"Expected at least one of `Bootstrap.ConfigRef` or `Bootstrap.Data` to be populated for MachinePool %q in namespace %q",
```
In v1alpha3 we have moved to `Bootstrap.DataSecretName` instead

vincepri commented on controllers/machinepool_controller_phases.go:
```diff
@@ -0,0 +1,232 @@
+		return errors.Errorf("retrieved empty data from bootstrap provider for MachinePool %q in namespace %q", mp.Name, mp.Namespace)
+	}
+
+	mp.Spec.Template.Spec.Bootstrap.Data = pointer.StringPtr(data)
```
We need to get the bootstrap data from the secret here, instead of the field in plaintext, which has been deprecated in v1alpha3

ncdc commented on controllers/machinepool_controller_noderef.go:
```diff
@@ -0,0 +1,167 @@
+func (r *MachinePoolReconciler) getNodeReferences(ctx context.Context, client client.Client, providerIDs []string) ([]apicorev1.ObjectReference, int, int, error) {
+	var ready, available int
+	nodeRefsMap := make(map[string]apicorev1.Node)
+	listOpt := metav1.ListOptions{}
```
I'm guessing you were copying this from MachineReconciler.getNodeReference, which isn't doing this correctly 😄. This is roughly how it should look:

```go
nodeList := apicorev1.NodeList{}
for {
  if err := c.List(ctx, &nodeList, client.Continue(nodeList.Continue)); err != nil {
    // check error
  }

  // do stuff

  if nodeList.Continue == "" {
    break
  }
}
```

ncdc replied:
(I filed https://github.com/kubernetes-sigs/cluster-api/issues/2166 for MachineReconciler)

ncdc commented on controllers/machinepool_controller_noderef.go:
```diff
@@ -0,0 +1,167 @@
+		}
+
+		for _, node := range nodeList.Items {
+			nodeRefsMap[node.Spec.ProviderID] = node
```
Same question as before about noderefutil.NewProviderID?

ncdc commented on controllers/machinepool_controller_noderef.go:
```diff
@@ -0,0 +1,167 @@
+		}
+	}
+
+	if nodeRefs == nil {
```
```suggestion
	if len(nodeRefs) == 0 {
```

ncdc commented on controllers/machinepool_controller_noderef.go:
```diff
@@ -0,0 +1,167 @@
+	return nodeRefs, available, ready, nil
+}
+
+func isReady(node *apicorev1.Node) bool {
```
```suggestion
func nodeIsReady(node *apicorev1.Node) bool {
```

ncdc commented on controllers/machinepool_controller_phases.go:
```diff
@@ -0,0 +1,232 @@
+	}
+
+	// Set the phase to "running" if there is a NodeRef field.
+	if mp.Status.InfrastructureReady && mp.Status.NodeRefs != nil && len(mp.Status.NodeRefs) == int(mp.Status.ReadyReplicas) {
```
```suggestion
	if mp.Status.InfrastructureReady && len(mp.Status.NodeRefs) == int(mp.Status.ReadyReplicas) {
```

ncdc commented on controllers/machinepool_controller_phases.go:
```diff
@@ -0,0 +1,232 @@
+		return nil, err
+	}
+	if errorReason != "" {
+		machinePoolStatusError := capierrors.MachinePoolStatusError(errorReason)
```
Can we rename to MachinePoolStatusFailure?

ncdc commented on controllers/machinepool_controller_phases.go:
```diff
@@ -0,0 +1,232 @@
+		}
+	}
+
+	// Set error reason and message, if any.
```
```suggestion
	// Set failure reason and message, if any.
```

ncdc commented on controllers/machinepool_controller_phases.go:
```diff
@@ -0,0 +1,232 @@
+	}
+
+	// Set error reason and message, if any.
+	errorReason, errorMessage, err := external.FailuresFrom(obj)
```
```suggestion
	failureReason, failureMessage, err := external.FailuresFrom(obj)
```

ncdc commented on controllers/machinepool_controller_phases.go:
```diff
@@ -0,0 +1,232 @@
+	if err != nil {
+		return nil, err
+	}
+	if errorReason != "" {
```
```suggestion
	if failureReason != "" {
```

ncdc commented on controllers/machinepool_controller_noderef.go:
```diff
@@ -0,0 +1,167 @@
+	return nil
+}
+
+func (r *MachinePoolReconciler) getNodeReferences(ctx context.Context, client client.Client, providerIDs []string) ([]apicorev1.ObjectReference, int, int, error) {
```
```suggestion
func (r *MachinePoolReconciler) getNodeReferences(ctx context.Context, c client.Client, providerIDs []string) ([]apicorev1.ObjectReference, int, int, error) {
```

ncdc commented on controllers/machinepool_controller_phases.go:
```diff
@@ -0,0 +1,232 @@
+	}
+
+	// Add watcher for external object, if there isn't one already.
+	_, loaded := r.externalWatchers.LoadOrStore(obj.GroupVersionKind().String(), struct{}{})
```
xref #2140 #2168 (no action required now, unless we solve those before this gets merged)

ncdc commented on controllers/machinepool_controller_phases.go:
```diff
@@ -0,0 +1,232 @@
+		machinePoolStatusError := capierrors.MachinePoolStatusError(errorReason)
+		mp.Status.FailureReason = &machinePoolStatusError
+	}
+	if errorMessage != "" {
```
```suggestion
	if failureMessage != "" {
```

ncdc commented on controllers/machinepool_controller_phases.go:
```diff
@@ -0,0 +1,232 @@
+	// Get Spec.ProviderIDs from the infrastructure provider.
+	if err := util.UnstructuredUnmarshalField(infraConfig, &mp.Spec.ProviderIDs, "spec", "providerIDs"); err != nil {
+		return errors.Wrapf(err, "failed to retrieve data from infrastructure provider for MachinePool %q in namespace %q", mp.Name, mp.Namespace)
+	} else if mp.Spec.ProviderIDs == nil {
```
```suggestion
	} else if len(mp.Spec.ProviderIDs) == 0 {
```

ncdc commented:
@juan-lee @CecileRobertMichon will you all have time over the next week or so to work on review comments & unit tests? Anything else we can do to help?

juan-lee commented:
> @juan-lee @CecileRobertMichon will you all have time over the next week or so to work on review comments & unit tests? Anything else we can do to help?

@ncdc I just addressed the comments sans validation and unit tests. I should have cycles today or tomorrow to get those wrapped up. Thanks for the review @ncdc @vincepri @CecileRobertMichon @JoelSpeed

ncdc commented on bootstrap/kubeadm/controllers/kubeadmconfig_controller.go:
```diff
@@ -535,6 +541,22 @@ func (r *KubeadmConfigReconciler) MachineToBootstrapMapFunc(o handler.MapObject)
+	if !ok {
+		return nil
+	}
+	if m.Spec.Template.Spec.Bootstrap.ConfigRef != nil && m.Spec.Template.Spec.Bootstrap.ConfigRef.GroupVersionKind() == bootstrapv1.GroupVersion.WithKind("KubeadmConfig") {
```
```suggestion
	if m.Spec.Template.Spec.Bootstrap.ConfigRef != nil && m.Spec.Template.Spec.Bootstrap.ConfigRef.GroupVersionKind().GroupKind() == bootstrapv1.GroupVersion.WithKind("KubeadmConfig").GroupKind() {
```

ncdc commented on controllers/machinepool_controller.go:
```diff
@@ -57,13 +76,181 @@ func (r *MachinePoolReconciler) SetupWithManager(mgr ctrl.Manager, options contr
+			// For additional cleanup logic use finalizers.
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
```
Let's log the error so we know where it came from

ncdc commented on controllers/machinepool_controller.go:
```diff
@@ -57,13 +76,181 @@ func (r *MachinePoolReconciler) SetupWithManager(mgr ctrl.Manager, options contr
+
+	cluster, err := util.GetClusterByName(ctx, r.Client, mp.ObjectMeta.Namespace, mp.Spec.ClusterName)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to get cluster %q for machine %q in namespace %q",
```
Let's log the error so we know where it came from

ncdc commented on controllers/machinepool_controller.go:
```diff
@@ -57,13 +76,181 @@ func (r *MachinePoolReconciler) SetupWithManager(mgr ctrl.Manager, options contr
 	r.config = mgr.GetConfig()
-
 	r.scheme = mgr.GetScheme()
 	return nil
```
Pending resolution of https://github.com/kubernetes-sigs/cluster-api/pull/2101/files#r371995294, may need to add a watch map function from Cluster -> MachinePool

detiber replied:
We should likely also port over the immutability check on `ClusterName` as well.

juan-lee replied:
The immutability check is part of a validation webhook @detiber? Would it be okay to enable the validation webhooks and immutability check in a follow-up PR?

juan-lee replied:
I'm not sure I fully understand what's required here. @vincepri @detiber could either of you clarify?

vincepri replied:
I'm not totally sure (the link to the discussion is not working), but do we need to receive reconcile requests here when an Cluster object is updated?

juan-lee replied:
I have not encountered the need for triggering when cluster is updated during my testing, but I might not be exercising the scenario ya'll have in mind. @vincepri @detiber @ncdc Could I get some clarification or can we address this in a subsequent PR?

ncdc commented on controllers/machinepool_controller.go:
```diff
@@ -57,13 +76,181 @@ func (r *MachinePoolReconciler) SetupWithManager(mgr ctrl.Manager, options contr
+}
+
+func (r *MachinePoolReconciler) reconcile(ctx context.Context, cluster *clusterv1.Cluster, mp *clusterv1.MachinePool) (ctrl.Result, error) {
+	// If the MachinePool belongs to a cluster, add an owner reference.
```
As discussed in #2101, we can probably omit the call to shouldAdopt and always ensure there is an owner ref to the cluster.

detiber commented on bootstrap/kubeadm/controllers/kubeadmconfig_controller.go:
```diff
@@ -535,6 +541,22 @@ func (r *KubeadmConfigReconciler) MachineToBootstrapMapFunc(o handler.MapObject)
+	if !ok {
+		return nil
+	}
+	if m.Spec.Template.Spec.Bootstrap.ConfigRef != nil && m.Spec.Template.Spec.Bootstrap.ConfigRef.GroupVersionKind().GroupKind() == bootstrapv1.GroupVersion.WithKind("KubeadmConfig").GroupKind() {
```
Based on other discussions we've had around handling/updating of versioning of objectReferences, I'm wondering if this should only compare group and kind, ignoring version?

@vincepri we'd also need to take that into account for other similar use cases such as the `MachineToBootstrapMapFunc` method above.

ncdc replied:
I had that in one of my previous comments. Must have gotten lost w/the most recent push.

vincepri replied:
The map functions I believe can only be registered and receive one specific version of an object, unless you meant something else?

ncdc replied:
Oh what, that change did make it in here!

detiber replied:
This is less around the registering of the types for the watching and more about ignoring the api version of the remote object reference on the MachinePool resource that is passed into the map function.

detiber commented on controllers/machinepool_controller.go:
```diff
@@ -57,13 +75,179 @@ func (r *MachinePoolReconciler) SetupWithManager(mgr ctrl.Manager, options contr
+	}
+
+	// If the MachinePool doesn't have a finalizer, add one.
+	if !util.Contains(mp.Finalizers, clusterv1.MachinePoolFinalizer) {
```
I wonder if we should start moving to use `AddFinalizerWithError` and `RemoveFinalizerWithError` from [sigs.k8s.io/controller-runtime/pkg/controller/controllerutil](https://godoc.org/sigs.k8s.io/controller-runtime/pkg/controller/controllerutil) for adding/removing finalizers?

ncdc replied:
Yeah, new issue please. And I think we can use the non-error version?

detiber replied:
A PR for the filed issue was already merged: https://github.com/kubernetes-sigs/cluster-api/pull/2205, so we should probably adopt that approach for this PR now as well.

ncdc commented on controllers/machinepool_controller_phases.go:
```diff
@@ -0,0 +1,260 @@
+	}
+
+	// Add ownerRef to object.
+	obj.SetOwnerReferences(util.EnsureOwnerRef(obj.GetOwnerReferences(), ownerRef))
```
@vincepri @detiber do you want a controller owner here?

detiber replied:
Yes, we should set a controllerref here similar to what we are doing in the MachineController's reconcileExternal: https://github.com/kubernetes-sigs/cluster-api/blob/4ec4ec416a91317f4e43cb2d885f592b776072c0/controllers/machine_controller_phases.go#L102-L105

ncdc commented on controllers/machinepool_controller_phases.go:
```diff
@@ -0,0 +1,260 @@
+	}
+
+	// If the bootstrap data is populated, set ready and return.
+	if m.Spec.Template.Spec.Bootstrap.Data != nil || m.Spec.Template.Spec.Bootstrap.DataSecretName != nil {
```
@vincepri do we want to be looking at .data?

vincepri replied:
We should only look at the secret

ncdc commented:
@juan-lee just checking in to see if you saw the most recent comments? Also, do you have an ETA for unit tests?

juan-lee commented:
> @juan-lee just checking in to see if you saw the most recent comments? Also, do you have an ETA for unit tests?

@ncdc I've addressed some of the recent comments yesterday. I'm nearly finished with unit tests. I should have them pushed later today or early tomorrow.

juan-lee commented:
@ncdc @vincepri @detiber All comments have been addressed (pending clarification on one) and validation/unit tests are done. Would love a final pass over the PR.

vincepri commented:
Taking a look today

vincepri commented on bootstrap/kubeadm/controllers/kubeadmconfig_controller.go:
```diff
@@ -554,6 +560,22 @@ func (r *KubeadmConfigReconciler) MachineToBootstrapMapFunc(o handler.MapObject)
+	if !ok {
+		return nil
+	}
+	if m.Spec.Template.Spec.Bootstrap.ConfigRef != nil && m.Spec.Template.Spec.Bootstrap.ConfigRef.GroupVersionKind().GroupKind() == bootstrapv1.GroupVersion.WithKind("KubeadmConfig").GroupKind() {
```
Nit: assign `m.Spec.Template.Spec.Bootstrap.ConfigRef` to a variable so it's a little more readable

vincepri commented on controllers/machinepool_controller.go:
```diff
@@ -57,13 +72,175 @@ func (r *MachinePoolReconciler) SetupWithManager(mgr ctrl.Manager, options contr
+
+	cluster, err := util.GetClusterByName(ctx, r.Client, mp.ObjectMeta.Namespace, mp.Spec.ClusterName)
+	if err != nil {
+		logger.Error(err, "Failed to get cluster for machinepool.")
```
```suggestion
		logger.Error(err, "Failed to get Cluster %s for MachinePool", mp.Spec.ClusterName)
```

vincepri commented on controllers/machinepool_controller.go:
```diff
@@ -57,13 +72,175 @@ func (r *MachinePoolReconciler) SetupWithManager(mgr ctrl.Manager, options contr
+		return err
+	}
+
+	if err := r.deleteRetiredNodes(ctx, clusterClient, machinepool.Status.NodeRefs, machinepool.Spec.ProviderIDs); err != nil {
```
Sidenote, should we rename the list to `ProviderIDList` or something similar?

vincepri commented on controllers/machinepool_controller_noderef.go:
```diff
@@ -0,0 +1,202 @@
+	}
+
+	// Check that the Machine doesn't already have a NodeRefs.
+	if len(machinepool.Status.NodeRefs) == int(machinepool.Status.Replicas) {
```
What if we add and remove a replica very quickly? Would that get us to have stale data?

juan-lee replied:
I changed this to be a bit more robust. Now it also checks if replicas == readyReplicas, with readyReplicas being reset if the infraProvider state changes.

vincepri commented on controllers/machinepool_controller_noderef.go:
```diff
@@ -0,0 +1,202 @@
+
+	// Check that the MachinePool has valid ProviderIDs.
+	if len(machinepool.Spec.ProviderIDs) == 0 {
+		logger.Info("MachinePool doesn't have valid ProviderIDs yet")
```
Should this be a V(2) or more? Also I'd `s/valid/any`

vincepri commented on controllers/machinepool_controller_noderef.go:
```diff
@@ -0,0 +1,202 @@
+		return err
+	}
+
+	err = r.deleteRetiredNodes(ctx, clusterClient, machinepool.Status.NodeRefs, machinepool.Spec.ProviderIDs)
```
if err := ...; err != nil

vincepri commented on controllers/machinepool_controller_noderef.go:
```diff
@@ -0,0 +1,202 @@
+		return err
+	}
+
+	// Get the Node reference.
```
```suggestion
	// Get the Node references.
```

vincepri commented on controllers/machinepool_controller_noderef.go:
```diff
@@ -0,0 +1,202 @@
+			return errors.Wrapf(&capierrors.RequeueAfterError{RequeueAfter: 10 * time.Second},
+				"cannot assign NodeRefs to MachinePool %q in namespace %q, no matching Nodes", machinepool.Name, machinepool.Namespace)
+		}
+		logger.Error(err, "Failed to assign NodeRef")
```
There is probably no need to log here, given that we're returning the error

vincepri commented on controllers/machinepool_controller_noderef.go:
```diff
@@ -0,0 +1,202 @@
+		}
+		logger.Error(err, "Failed to assign NodeRef")
+		r.recorder.Event(machinepool, apicorev1.EventTypeWarning, "FailedSetNodeRef", err.Error())
+		return err
```
Wrap this error

vincepri commented on controllers/machinepool_controller_noderef.go:
```diff
@@ -0,0 +1,202 @@
+	machinepool.Status.AvailableReplicas = int32(nodeRefsResult.available)
+	machinepool.Status.ReadyReplicas = int32(nodeRefsResult.ready)
+	machinepool.Status.NodeRefs = nodeRefsResult.references
+	logger.Info("Set MachinePools's NodeRefs", "noderefs", machinepool.Status.NodeRefs)
```
Leave space above?

vincepri commented on controllers/machinepool_controller_noderef.go:
```diff
@@ -0,0 +1,202 @@
+	r.recorder.Event(machinepool, apicorev1.EventTypeNormal, "SuccessfulSetNodeRef", fmt.Sprintf("%+v", machinepool.Status.NodeRefs))
+
+	if machinepool.Status.ReadyReplicas == 0 || len(nodeRefsResult.references) != int(machinepool.Status.ReadyReplicas) {
+		return errors.Wrapf(&capierrors.RequeueAfterError{RequeueAfter: 10 * time.Second},
```
Is this a little bit aggressive?

vincepri commented on controllers/machinepool_controller_noderef.go:
```diff
@@ -0,0 +1,202 @@
+// deleteRetiredNodes deletes nodes that don't have a corresponding ProviderID in Spec.ProviderIDs.
+// A MachinePool infrastucture provider indicates an instance in the set has been deleted by
+// removing its ProviderID from the slice.
+func (r *MachinePoolReconciler) deleteRetiredNodes(ctx context.Context, client client.Client, nodeRefs []apicorev1.ObjectReference, providerIDs []string) error {
```
Is `client` a remote client? If so, I'd rename it

vincepri commented on controllers/machinepool_controller_noderef.go:
```diff
@@ -0,0 +1,202 @@
+	nodeRefsMap := make(map[string]*apicorev1.Node, len(nodeRefs))
+	for _, nodeRef := range nodeRefs {
+		node := &corev1.Node{}
+		err := client.Get(ctx, types.NamespacedName{Name: nodeRef.Name}, node)
```
if err := ...; err != nil

vincepri commented on controllers/machinepool_controller_noderef.go:
```diff
@@ -0,0 +1,202 @@
+		node := &corev1.Node{}
+		err := client.Get(ctx, types.NamespacedName{Name: nodeRef.Name}, node)
+		if err != nil {
+			return err
```
Wrap this error?

vincepri commented on controllers/machinepool_controller_noderef.go:
```diff
@@ -0,0 +1,202 @@
+
+		nodeProviderID, err := noderefutil.NewProviderID(node.Spec.ProviderID)
+		if err != nil {
+			logger.Error(err, "Failed to parse ProviderID", "node", node.Name)
```
This seems more of a warning/info? We continue the iteration, and I don't think users can take any actions

vincepri commented on controllers/machinepool_controller_noderef.go:
```diff
@@ -0,0 +1,202 @@
+	for _, providerID := range providerIDs {
+		pid, err := noderefutil.NewProviderID(providerID)
+		if err != nil {
+			logger.Error(err, "Failed to parse", "providerID", providerID)
```
Same as above

vincepri commented on controllers/machinepool_controller_noderef.go:
```diff
@@ -0,0 +1,202 @@
+		delete(nodeRefsMap, pid.ID())
+	}
+	for _, node := range nodeRefsMap {
+		err := client.Delete(ctx, node)
```
if err := ...; err != nil

vincepri commented on controllers/machinepool_controller_noderef.go:
```diff
@@ -0,0 +1,202 @@
+	for _, node := range nodeRefsMap {
+		err := client.Delete(ctx, node)
+		if err != nil {
+			return err
```
Wrap this error?

vincepri commented on controllers/machinepool_controller_noderef.go:
```diff
@@ -0,0 +1,202 @@
+	nodeRefsMap := make(map[string]apicorev1.Node)
+	nodeList := apicorev1.NodeList{}
+	for {
+		err := c.List(ctx, &nodeList, client.Continue(nodeList.Continue))
```
if err :=; err != nil

vincepri commented on controllers/machinepool_controller_noderef.go:
```diff
@@ -0,0 +1,202 @@
+	for {
+		err := c.List(ctx, &nodeList, client.Continue(nodeList.Continue))
+		if err != nil {
+			return getNodeReferencesResult{}, err
```
wrap the error?

vincepri commented on controllers/machinepool_controller_noderef.go:
```diff
@@ -0,0 +1,202 @@
+		for _, node := range nodeList.Items {
+			nodeProviderID, err := noderefutil.NewProviderID(node.Spec.ProviderID)
+			if err != nil {
+				logger.Error(err, "Failed to parse ProviderID", "node", node.Name)
```
Same as above, more a warning/info?

vincepri commented on controllers/machinepool_controller_noderef.go:
```diff
@@ -0,0 +1,202 @@
+	for _, providerID := range providerIDs {
+		pid, err := noderefutil.NewProviderID(providerID)
+		if err != nil {
+			logger.Error(err, "Failed to parse", "providerID", providerID)
```
Same as above, more a warning/info?

vincepri commented on controllers/machinepool_controller_noderef.go:
```diff
@@ -0,0 +1,202 @@
+		if node, ok := nodeRefsMap[pid.ID()]; ok {
+			available++
+			nodeReady := nodeIsReady(&node)
+			if nodeReady {
```
```suggestion
			if nodeIsReady(&node) {
```

vincepri commented on controllers/machinepool_controller_noderef.go:
```diff
@@ -0,0 +1,202 @@
+	}
+
+	if len(nodeRefs) == 0 {
+		return getNodeReferencesResult{}, ErrNodeNotFound
```
Where is this error defined? I'd have our own error in this file, but with something more specific to MachinePool.

Example: `ErrNoAvailableNodes`

juan-lee commented:
/test pull-cluster-api-capd-e2e

vincepri commented:
/milestone v0.3.0-rc.1

juan-lee commented:
/assign @davidewatson

juan-lee commented:
/retest

vincepri commented:
/lgtm
